substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/jfr/JfrBufferList.java

```diff
@@ -0,0 +1,166 @@
+ * Copyright (c) 2022, 2022, Red Hat Inc. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.  Oracle designates this
+ * particular file as subject to the "Classpath" exception as provided
+ * by Oracle in the LICENSE file that accompanied this code.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package com.oracle.svm.core.jfr;
+
+import org.graalvm.nativeimage.ImageSingletons;
+import org.graalvm.nativeimage.Platform;
+import org.graalvm.nativeimage.Platforms;
+import org.graalvm.nativeimage.impl.UnmanagedMemorySupport;
+import org.graalvm.word.WordFactory;
+
+import com.oracle.svm.core.Uninterruptible;
+import com.oracle.svm.core.thread.JavaSpinLockUtils;
+import com.oracle.svm.core.util.VMError;
+
+import jdk.internal.misc.Unsafe;
+
+/**
+ * Singly linked list that stores {@link JfrBuffer}s. Multiple instances of this data structure are
+ * used to keep track of the global and the various thread-local buffers. When entering a safepoint,
+ * it is guaranteed that none of the blocked Java threads holds the list's lock.
+ *
+ * The following invariants are crucial if the list is used for thread-local buffers:
+ * <ul>
+ * <li>Each thread shall only add one node to the list.</li>
+ * <li>Only threads that hold the {@link JfrChunkWriter#lock()} may iterate or remove nodes from the
+ * list.</li>
+ * </ul>
+ */
+public class JfrBufferList {
+    private static final Unsafe U = Unsafe.getUnsafe();
+    private static final long LOCK_OFFSET = U.objectFieldOffset(JfrBufferList.class, "lock");
+
+    @SuppressWarnings("unused") private volatile int lock;
+    private JfrBufferNode head;
+
+    @Platforms(Platform.HOSTED_ONLY.class)
+    public JfrBufferList() {
+    }
+
+    /**
+     * This method is called after the {@link JfrBuffer}s were already flushed and freed, but there
+     * may still be nodes in the list.
+     */
+    public void teardown() {
+        JfrBufferNode node = head;
+        while (node.isNonNull()) {
+            assert node.getBuffer().isNull();
```

Comment:
Is this assertion and the Javadoc above always true? When `destroyJfr` is called and the java buffer node list is being torn down, when are the JfrBuffers freed before reaching here?

Reply:
You are right, this is one of the open topics. I need to think about this a bit more because I don't have a good solution yet.